addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java

```diff
@@ -0,0 +1,1299 @@
+import org.openhab.binding.spotify.SpotifyBindingConstants;
+import org.openhab.binding.spotify.handler.SpotifyHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * The {@link SpotifySession} maintains an active session with Spotify Web API.
+ * The refreshToken is used to retrieve an accessToken which is used through its validity time.
+ * The class also provides wrappers for various API calls.
+ *
+ * @author Andreas Stenlund - Initial contribution
+ */
+public class SpotifySession implements Runnable {
+
+    private final Logger logger = LoggerFactory.getLogger(SpotifySession.class);
+
+    private final int HTTP_CLIENT_RETRY_COUNT = 5;
+    private final int HTTP_CLIENT_TIMEOUT = 30;
+    private final int ASYNC_RESULT_WAIT_TIME = 10; // ms to
+
+    // Instantiate and configure the SslContextFactory
+    private SslContextFactory sslContextFactory = new SslContextFactory();
+
+    // Instantiate HttpClient with the SslContextFactory
+    private HttpClient httpClient = new HttpClient(sslContextFactory);
+
+    private String clientId = null;
+    private String clientSecret = null;
+    private String refreshToken = null;
+    private String accessToken = null;
+    private int tokenValidity = 3600;
+
+    private SpotifyHandler spotifyPlayer = null;
+    private final ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
+    private final ExecutorService syncExecutorService = Executors.newFixedThreadPool(1);
+    private final ExecutorService asyncExecutorService = Executors.newFixedThreadPool(1);
+
+    @SuppressWarnings({ "rawtypes" })
+    private ScheduledFuture future = null;
+
+    private SpotifySession(SpotifyHandler player, String clientId, String clientSecret, String refreshToken) {
+        this.spotifyPlayer = player;
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.refreshToken = refreshToken;
+
+        httpClient.setFollowRedirects(true);
+        if (!httpClient.isStarted()) {
+            try {
+                httpClient.start();
+            } catch (Exception e) {
+                logger.error("Error starting HttpClient", e);
+            }
+
+        }
+    }
+
+    public static SpotifySession getInstance(SpotifyHandler player, String clientId, String clientSecret,
+            String refreshToken) {
+
+        SpotifySession session = new SpotifySession(player, clientId, clientSecret, refreshToken);
+        return session;
+    }
+
+    /**
+     *
+     */
+    public void dispose() {
+        if (future != null) {
+            future.cancel(true);
+        }
+
+        scheduledExecutorService.shutdownNow();
+        syncExecutorService.shutdownNow();
+
+        // TODO: Not really need? Causing bad exceptions when refreshing SpotifySession in Controller check later....
```

Comment:
There is a TODO, what should be done and can it be completed before merging the PR?